File: kernel/drivers/src/vga.zig
const c = @cImport({
    @cInclude("drivers/vga.h");
});

const VGA_WIDTH: usize = c.VGA_WIDTH;
const VGA_HEIGHT: usize = c.VGA_HEIGHT;

const VGA_MEMORY = 0xC00B8000;

const VgaColour = enum(u8) {
    VGA_COLOR_BLACK = 0,
    VGA_COLOR_BLUE = 1,
    VGA_COLOR_GREEN = 2,
    VGA_COLOR_CYAN = 3,
    VGA_COLOR_RED = 4,
    VGA_COLOR_MAGENTA = 5,
    VGA_COLOR_BROWN = 6,
    VGA_COLOR_LIGHT_GREY = 7,
    VGA_COLOR_DARK_GREY = 8,
    VGA_COLOR_LIGHT_BLUE = 9,
    VGA_COLOR_LIGHT_GREEN = 10,
    VGA_COLOR_LIGHT_CYAN = 11,
    VGA_COLOR_LIGHT_RED = 12,
    VGA_COLOR_LIGHT_MAGENTA = 13,
    VGA_COLOR_LIGHT_BROWN = 14,
    VGA_COLOR_WHITE = 15,
};

fn vgaEntryColour(fg: VgaColour, bg: VgaColour) u8 {
    return @intFromEnum(fg) | (@intFromEnum(bg) << 4);
}

fn vgaEntry(ch: u8, colour: u8) u16 {
    return @as(u16, ch) | (@as(u16, colour) << 8);
}

pub const Writer = struct {
    column: usize,
    row: usize,
    colour: u8,
    buffer: [*]volatile u16,

    pub fn new() Writer {
        return Writer{
            .column = 0,
            .row = 0,
            .colour = vgaEntryColour(.VGA_COLOR_WHITE, .VGA_COLOR_BLACK),
            .buffer = @ptrFromInt(VGA_MEMORY),
        };
    }

    pub fn init(self: *Writer) void {
        self.clearScreen();
    }

    pub fn writeString(self: *Writer, str: [*:0]const u8) void {
        var i: usize = 0;

        while (str[i] != 0) : (i += 1) {
            self.writeChar(str[i]);
        }
    }

    pub fn writeChar(self: *Writer, ch: u8) void {
        if (ch == '\n') {
            self.newLine();
            return;
        }

        const index = self.row * VGA_WIDTH + self.column;
        self.buffer[index] = vgaEntry(ch, self.colour);

        self.column += 1;
        if (self.column >= VGA_WIDTH) {
            self.newLine();
        }
    }

    pub fn clearScreen(self: *Writer) void {
        for (0..VGA_HEIGHT) |row| {
            for (0..VGA_WIDTH) |col| {
                const index = (row * VGA_WIDTH) + col;
                self.buffer[index] = vgaEntry(' ', self.colour);
            }
        }

        self.column = 0;
        self.row = 0;
    }

    pub fn clearChar(self: *Writer) void {
        if (self.column > 0) {
            self.column -= 1;
            const index = self.row * VGA_WIDTH + self.column;
            self.buffer[index] = vgaEntry(' ', self.colour);
        }
    }

    fn newLine(self: *Writer) void {
        self.column = 0;

        if (self.row < VGA_HEIGHT - 1) {
            self.row += 1;
            return;
        }

        self.scroll();
    }

    fn scroll(self: *Writer) void {
        for (1..VGA_HEIGHT) |row| {
            for (0..VGA_WIDTH) |col| {
                const src_index = row * VGA_WIDTH + col;
                const dst_index = (row - 1) * VGA_WIDTH + col;
                self.buffer[dst_index] = self.buffer[src_index];
            }
        }

        const blank: u16 = vgaEntry(' ', self.colour);
        for (0..VGA_WIDTH) |x| {
            const index = (VGA_HEIGHT - 1) * VGA_WIDTH + x;
            self.buffer[index] = blank;
        }
    }
};

var WRITER = Writer.new();

pub export fn vga_init() void {
    WRITER.init();
}

pub export fn vga_putchar(ch: u8) void {
    WRITER.writeChar(ch);
}

pub export fn vga_puts(str: [*:0]const u8) void {
    WRITER.writeString(str);
}

pub export fn vga_clear() void {
    WRITER.clearScreen();
}

pub export fn vga_delete() void {
    WRITER.clearChar();
}

File: kernel/drivers/src/vga.c
#include <drivers/serial.h>
#include <drivers/vga.h>
#include <ferrite/types.h>

#define VGA_ENTRY_COLOUR(fg, bg) ((u8)(fg) | ((u8)(bg) << 4))
#define VGA_ENTRY(c, colour) ((u16)(c) | ((u16)(colour) << 8))

#define WRITER_NEW                                                  \
    { .column = 0,                                                  \
      .row = 0,                                                     \
      .colour = VGA_ENTRY_COLOUR(VGA_COLOR_WHITE, VGA_COLOR_BLACK), \
      .buffer = (volatile u16*)VGA_MEMORY }

static Writer WRITER = WRITER_NEW;

/* Private */

static void writer_scroll(Writer* self)
{
    for (size_t y = 1; y < VGA_HEIGHT; y++) {
        for (size_t x = 0; x < VGA_WIDTH; x++) {
            size_t src_index = y * VGA_WIDTH + x;
            size_t dst_index = (y - 1) * VGA_WIDTH + x;
            self->buffer[dst_index] = self->buffer[src_index];
        }
    }

    u16 blank = VGA_ENTRY(' ', self->colour);
    for (size_t x = 0; x < VGA_WIDTH; x++) {
        size_t index = (VGA_HEIGHT - 1) * VGA_WIDTH + x;
        self->buffer[index] = blank;
    }
}

static void writer_new_line(Writer* self)
{
    self->column = 0;
    if (self->row < VGA_HEIGHT - 1) {
        self->row += 1;
        return;
    }

    writer_scroll(self);
}

static void writer_write_char(Writer* self, u8 c)
{
#if defined(__print_serial)
    serial_write_byte(c);
#endif

    if (c == '\n') {
        writer_new_line(self);
        return;
    }

    size_t index = self->row * VGA_WIDTH + self->column;
    self->buffer[index] = VGA_ENTRY(c, self->colour);
    self->column += 1;

    if (self->column >= VGA_WIDTH) {
        writer_new_line(self);
    }
}

static void writer_write_string(Writer* self, char const* str)
{
    for (size_t i = 0; str[i] != '\0'; i++) {
        writer_write_char(self, str[i]);
    }
}

static void writer_clear_screen(Writer* self)
{
    for (size_t r = 0; r < VGA_HEIGHT; r++) {
        for (size_t col = 0; col < VGA_WIDTH; col++) {
            size_t index = r * VGA_WIDTH + col;
            self->buffer[index] = VGA_ENTRY(' ', self->colour);
        }
    }
    self->column = 0;
    self->row = 0;
}

static void writer_clear_char(Writer* self)
{
    if (self->column > 0) {
        self->column -= 1;
        size_t index = self->row * VGA_WIDTH + self->column;
        self->buffer[index] = VGA_ENTRY(' ', self->colour);
    }
}

static void writer_init(Writer* self)
{
    for (size_t y = 0; y < VGA_HEIGHT; y++) {
        for (size_t x = 0; x < VGA_WIDTH; x++) {
            size_t index = y * VGA_WIDTH + x;
            self->buffer[index] = VGA_ENTRY(' ', self->colour);
        }
    }
}

/* Public */

void vga_init(void) { writer_init(&WRITER); }

void vga_putchar(u8 c) { writer_write_char(&WRITER, c); }

void vga_puts(char const* str) { writer_write_string(&WRITER, str); }

void vga_clear(void) { writer_clear_screen(&WRITER); }

void vga_delete(void) { writer_clear_char(&WRITER); }
